buildenv/jenkins/jobs/infrastructure/Cleanup-Nodes

```diff
@@ -281,13 +297,33 @@ def sanitize_node(nodeName) {
         println(e.getMessage())
     }
 
+    //reconnect slave
     timeout(time: RECONNECT_TIMEOUT.toInteger(), unit: 'MINUTES') {
-        println("\t ${nodeName}: Reconnecting...")
+        println("\t ${nodeName}: Disconnecting...")
+        waitUntil {
+            // wait for Jenkins to disconnect the slave after kill -9 or reboot
+            return (workingComputer.getNode().getRootPath() == null)
+        }
         workingComputer.disconnect(null)
         workingComputer.waitUntilOffline()
-        workingComputer.connect(false)
-        workingComputer.setTemporarilyOffline(false, null)
-        workingComputer.waitUntilOnline()
-        println("${nodeName} is back online: ${workingComputer.isOnline()}")
+
+        println("\t ${nodeName}: Connecting...")
+        def reconnected = false
+        retry(2) {
+            if (reconnected) {
+                sleep(time: SLEEP_TIME.toInteger(), unit: 'SECONDS')
+            }
+            reconnected = true
+
+            workingComputer.connect(false)
+            waitUntil {
+                // wait for slave agent to relaunch
+                return (workingComputer.getChannel() != null)
+            }
+            workingComputer.setTemporarilyOffline(false, null)
+            workingComputer.waitUntilOnline()
```

Comment:
What is the difference between `waitUntilOnline()` and `waitUntil { getChannel() != null }` ?

Reply:
[waitUntilOnline()](https://javadoc.jenkins-ci.org/hudson/model/Computer.html#waitUntilOnline--) Blocks until the node becomes online. 
`connect()` returns immediately ( `isOnline()` returns true), but the relaunch operation happens asynchronously. `waitUntil` waits on the relaunch operation.

Reply:
I'm still confused. It looks to me that these do the same thing
```
waitUntil {
    // wait for slave agent to relaunch
    return (workingComputer.getChannel() != null)
}
```
```
workingComputer.waitUntilOnline()
```